webapp/src/ts/services/transitions/create-user-for-contacts.transition.ts

```diff
@@ -0,0 +1,173 @@
+  filter(docs: Doc[]) {
+    return !!docs.filter(doc => doc.type === 'data_record').length;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to run the transition over
+   * @returns {Promise<Array<Doc>>} - updated docs (may include additional docs)
+   */
+  async run(docs: Doc[]) {
+    const originalContact = await this.createUserForContactsService.getUserContact();
+    if (!originalContact) {
+      return docs;
+    }
+
+    if (this.createUserForContactsService.isReplaced(originalContact)) {
+      this.reparentReports(docs, originalContact);
+    }
+    const userReplaceDoc = this.getUserReplaceDoc(docs);
+    if (userReplaceDoc) {
+      return this.replaceUser(docs, userReplaceDoc, originalContact);
+    }
+
+    return docs;
+  }
+
+  private getTransitionConfig(settings) {
+    return settings[this.name];
+  }
+
+  private getReplaceForms(settings) {
+    return this.getTransitionConfig(settings)?.[this.REPLACE_PROPERTY] || [];
+  }
+
+  private getUserReplaceDoc(docs: Doc[]) {
+    return docs.find(doc => this.replaceForms.includes(doc.form)) as UserReplaceDoc;
+  }
+
+  private async replaceUser(docs: Doc[], userReplaceDoc: UserReplaceDoc, originalContact: Doc) {
+    const { replacement_contact_id } = userReplaceDoc.fields;
+    if (!replacement_contact_id) {
+      throw new Error('The form for replacing a user must include a replacement_contact_id field ' +
+        'containing the id of the new contact.');
+    }
+
+    const originalContactId = userReplaceDoc.contact._id;
+
+    const isOriginalContact = () => originalContact._id === originalContactId;
+    const isReplacedContact = () => this.createUserForContactsService.isReplaced(originalContact) &&
+      this.createUserForContactsService.getReplacedBy(originalContact) === originalContactId;
+    if (!isOriginalContact() && !isReplacedContact()) {
+      throw new Error('Only the contact associated with the currently logged in user can be replaced.');
+    }
+    const newContact = await this.getNewContact(docs, replacement_contact_id);
+    if (!newContact) {
+      throw new Error(`The new contact could not be found [${replacement_contact_id}].`);
+    }
+    this.createUserForContactsService.setReplaced(originalContact, newContact);
+    const updatedDocs = [...docs, originalContact];
+
+    const parentPlace = await this.getParentDoc(newContact);
+    if (parentPlace?.contact?._id === originalContactId) {
+      parentPlace.contact = this.extractLineageService.extract(newContact);
+      updatedDocs.push(parentPlace);
+    }
+
+    return updatedDocs;
+  }
+
+  private async getParentDoc(doc: ContactDoc) {
+    return this.dbService
+      .get()
+      .get(doc.parent._id)
+      .catch(err => {
+        if (err.status === 404) {
+          return;
+        }
+        throw err;
+      }) as ContactDoc;
+  }
+
+  private async getNewContact(docs: Doc[], newContactId: string): Promise<ContactDoc> {
+    const newContact = docs.find(doc => doc._id === newContactId);
```

Comment:
Hmm, this does not really make sense to me. I am all for being careful about making assumptions (especially when it comes to data coming from the DB), but unless I am missing something here, the only way `docs` could contain a `null` value would be if there was an actual bug _in the code somewhere._  In that case, it seems better to just blow up here with an error than silently suppress the issue.

I will add a null-check for all the docs in `TransitionsService.applyTransitions` method since that seems like the logical abstraction point where external "systems" would interact.  Let me know if you would prefer to null check `docs`at  the lower levels as well (either in the `run` function or all the way down here).

Reply:
I agree that it's good to blow up when an obvious error happened somewhere (another transition pushed null in the docs list for example or something equally dumb). 

I generally think that crashing as a side-effect in an unconnected place is generally not great. Suppose you had a bug in a form, and that form was in production for 1 month. 6 months later, Yuvraj makes an analysis of feedback docs and sees 1000 feedback docs, generated because errors were thrown in the `create-user-for-contact` transition (`can't access property _id of undefined`). How long and how possible it would be for someone to connect the dots with the broken form? 
I'm all for throwing errors when input is bad! but not randomly so that it produces unhelpful errors.